File: web/php/database/DatabaseAdapter.php
<?php

require "Key.php";
require "Insert.php";
require "Column.php";
require "Database.php";

class DatabaseAdapter {

    public function __construct() {
        Database::connect();
    }

    /**
     * Create table in database
     *
     * @param String $name Name of the table to be created in the database specified in Database.php
     * @param Column ...$row Rows to be added to the table
     */

    public function createTable($name, ...$row) {
        $stringBuilder = "";
        $stringBuilder .= "CREATE TABLE IF NOT EXISTS {$name} (";

        $length = sizeof($row);

        foreach ($row as $rows) {
            $stringBuilder .= $rows->getName() . " ";
            $separator = "";
            if ($length == 1) {
                $separator = ", ";
                $length--;
            }
            switch ($rows->getType()) {
                case "VARCHAR":
                    $stringBuilder .= "varchar(250)".$separator;
                    break;
                case "INTEGER":
                    $stringBuilder .= "bigint(250)".$separator;
                    break;
                case "TEXT":
                    $stringBuilder .= "text".$separator;
                    break;
                case "DOUBLE":
                    $stringBuilder .= "double(250, 2)".$separator;
                    break;
            }
        }
        $stringBuilder .= ");";
        $this->executeCommand($stringBuilder);
    }

    /**
     * @param String $query Valid SQL query string
     * @return mixed 0 if not successful, default return value of mysqli_query() if successful
     */
    public function executeCommand($query) {
        try {
            return mysqli_query(Database::getConnection(), $query);
        }
        catch (mysqli_sql_exception $exception) {
            return 0;
        }
    }

    /**
     * @param $tablename
     * @param $row
     * @param Key ...$keys
     *
     * @return String|int Returns the searched string if successful, if not, it returns 0
     */
    public function getStringFromTable($tablename, $row, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT {$row} FROM {$tablename} ";
        $length = sizeof($keys);
        if($length != 0) {
            foreach($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND ";
                    $length--;
                }
            }
        }
        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);
            while($returnRow = mysqli_fetch_assoc($result)) {
                return $returnRow[$row];
            }
        } catch (mysqli_sql_exception $exception) {}
        return 0;
    }

    /**
     * @param String $tablename
     * @param String $row
     * @param Key ...$keys
     *
     * @return array|int Returns the searched strings if successful, if not, it returns 0
     */
    public function getStringListFromTable($tablename, $row, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT {$row} FROM {$tablename} ";
        $length = sizeof($keys);
        if($length != 0) {
            foreach($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND ";
                    $length--;
                }
            }
        }
        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);

            $results = array();
            while($returnRow = mysqli_fetch_assoc($result)) {
                array_push($results, $returnRow[$row]);
            }
            return $results;
        } catch (mysqli_sql_exception $exception) {}
        return 0;
    }

    /**
     * @param String $tablename
     * @param Key ...$keys
     * @return array|null Array with associative arrays of the rows
     */
    public function getStringsFromTable($tablename, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT * FROM {$tablename} ";
        $length = sizeof($keys);
        if($length != 0) {
            foreach ($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND ";
                    $length--;
                }
            }
        }
        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);
            while($row = $result->fetch_assoc()) {
                return $row;
            }

        } catch (mysqli_sql_exception $exception) {

        }
        return null;
    }

    /**
     * @param String $tablename
     * @param String $orderByColumn
     * @return array|null Array with associative arrays of the rows
     */
    public function getAllStringsFromTable($tablename, $orderByColumn = null) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT * FROM {$tablename}";

        if($orderByColumn != null) {
            $stringBuilder .= " ORDER BY {$orderByColumn} ASC";
        }

        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);
            while($rows = $result->fetch_all(MYSQLI_ASSOC)) {
                return $rows;
            }

        } catch (mysqli_sql_exception $exception) {

        }
        return null;
    }

    /**
     * @param $tablename
     * @param $row
     * @param Key ...$keys
     *
     * @return null|int Returns an int if successful, if not, it returns null
     */
    public function getIntegerFromTable($tablename, $row, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT {$row} FROM {$tablename} ";
        $length = sizeof($keys);
        if($length != 0) {
            foreach($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND ";
                    $length--;
                }
            }
        }
        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);
            while($returnRow = mysqli_fetch_assoc($result)) {
                return (int)$returnRow[$row];
            }
        } catch (mysqli_sql_exception $exception) {}
        return null;
    }


    /**
     * @param String $tablename
     * @param Insert ...$inserts
     * @return bool|mysqli_result raw mysqli_query() return value
     */
    public function insertIntoTable($tablename, ...$inserts) {
        $stringBuilder = "";
        $stringBuilder .= "INSERT INTO {$tablename} (";

        $length = sizeof($inserts);

        foreach ($inserts as $insert) {
            if($length != 1) {
                $stringBuilder .= $insert->getColumn() . ", ";
                $length--;
            } else {
                $stringBuilder .= $insert->getColumn();
            }
        }

        $stringBuilder .= ") VALUES (";
        $length = sizeof($inserts);

        foreach ($inserts as $insert) {
            if($length != 1) {
                $stringBuilder .= "'" . $insert->getValue() . "', ";
                $length--;
            } else {
                $stringBuilder .= "'" . $insert->getValue() . "'";
            }
        }
        $stringBuilder .= ");";
        return mysqli_query(Database::getConnection(), $stringBuilder);
    }

    /**
     * @param String $tablename
     * @param String $row
     * @param String $newValue
     * @param Key ...$keys
     * @return mixed mysqli_query() return value
     */
    public function updateValue($tablename, $row, $newValue, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "UPDATE {$tablename} SET {$row} = '{$newValue}' ";

        $length = sizeof($keys);
        if($length != 0) {
            foreach ($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND";
                    $length--;
                }
            }
        }
        return $this->executeCommand($stringBuilder);
    }

    /**
     * @param $tablename
     * @param Key $key
     * @param Insert ...$inserts
     * @return mixed mysqli_query() return value
     */
    public function updateValues($tablename, $key, ...$inserts) {
        $stringBuilder = "";
        $stringBuilder .= "UPDATE {$tablename} SET ";

        $length = sizeof($inserts);
        foreach ($inserts as $insert) {
            if($length == 1) {
                $stringBuilder .= $insert->getColumn() . " = '" . $insert->getValue() . "' ";
            } else {
                $stringBuilder .= $insert->getColumn() . " = '" . $insert->getValue() . "', ";
                $length--;
            }
        }
        $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";

        return $this->executeCommand($stringBuilder);
    }

    /**
     * @param $tablename
     * @param Key ...$keys
     * @return true|false
     */
    public function containsEntry($tablename, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "SELECT * FROM {$tablename} ";

        $length = sizeof($keys);

        if($length != 0) {
            foreach ($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= "WHERE " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND ";
                    $length--;
                }
            }
        }
        try {
            $result = mysqli_query(Database::getConnection(), $stringBuilder);
            if(mysqli_num_rows($result) == 0) {
                return false;
            } else
                return true;
        } catch (mysqli_sql_exception $exception) {

        }
        return false;
    }

    /**
     * @param String $tablename
     * @param Key ...$keys
     * @return mixed mysqli_query() return value
     */
    public function deleteFromTable($tablename, ...$keys) {
        $stringBuilder = "";
        $stringBuilder .= "DELETE FROM {$tablename} WHERE";

        $length = sizeof($keys);
        if($length != 0) {
            foreach ($keys as $key) {
                if($length == 1) {
                    $stringBuilder .= " " . $key->getColumn() . " = '" . $key->getKeyWord() . "'";
                } else {
                    $stringBuilder .= " " . $key->getColumn() . " = '" . $key->getKeyWord() . "' AND";
                    $length--;
                }
            }
        }
        return $this->executeCommand($stringBuilder);
    }

}

File: server/src/main/java/com/voxcrafterlp/monitoring/server/database/DatabaseAdapter.java
package com.voxcrafterlp.monitoring.server.database;

import com.google.common.collect.Lists;
import com.voxcrafterlp.monitoring.server.Application;
import com.voxcrafterlp.monitoring.server.enums.RowType;
import com.voxcrafterlp.monitoring.server.objects.Insert;
import com.voxcrafterlp.monitoring.server.objects.Key;
import com.voxcrafterlp.monitoring.server.objects.Row;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

/**
 * This file was created by VoxCrafter_LP!
 * Date: 29.11.2020
 * Time: 16:00
 * Project: Remote-Monitoring
 */

public class DatabaseAdapter {

    private final Database database;

    public DatabaseAdapter() {
        this.database = Application.getInstance().getDatabase();
    }

    /**
     * Create table in database
     * @param name Name of table
     * @param row Requires a row array
     */
    public void createTable(String name, Row... row) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("CREATE TABLE IF NOT EXISTS "+name+" (");

        int length = row.length;

        for(Row rows : row) {
            stringBuilder.append(rows.getName() + " ");
            if(length == 1) {
                if(rows.getType() == RowType.VARCHAR) {
                    stringBuilder.append("varchar(250)");
                }
                if(rows.getType() == RowType.INTEGER) {
                    stringBuilder.append("bigint(250)");
                }
                if(rows.getType() == RowType.TEXT) {
                    stringBuilder.append("text");
                }
                if(rows.getType() == RowType.MEDIUMTEXT) {
                    stringBuilder.append("mediumtext");
                }
                if(rows.getType() == RowType.LONGTEXT) {
                    stringBuilder.append("longtext");
                }
                if(rows.getType() == RowType.DOUBLE) {
                    stringBuilder.append("double");
                }
            } else {
                if(rows.getType() == RowType.VARCHAR) {
                    stringBuilder.append("varchar(250), ");
                }
                if(rows.getType() == RowType.INTEGER) {
                    stringBuilder.append("bigint(250), ");
                }
                if(rows.getType() == RowType.TEXT) {
                    stringBuilder.append("text, ");
                }
                if(rows.getType() == RowType.MEDIUMTEXT) {
                    stringBuilder.append("mediumtext, ");
                }
                if(rows.getType() == RowType.LONGTEXT) {
                    stringBuilder.append("longtext, ");
                }
                if(rows.getType() == RowType.DOUBLE) {
                    stringBuilder.append("double, ");
                }
                length--;
            }
        }
        stringBuilder.append(");");
        executeCommand(stringBuilder.toString());
    }

    /**
     * Sends a command to the database
     * @param query Requires a query String
     */
    public void executeCommand(String query) {
        try {
            database.getConnection().prepareStatement(query).executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    /**
     * Gets a string from a table
     * @param tablename Requires a table name
     * @param row Requires the row of target string
     * @param key (Optional) key
     * @return Returns the string from the table
     */
    public String getStringFromTable(String tablename, String row, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("SELECT " + row + " FROM " + tablename + " ");
        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        try {
            ResultSet resultSet = database.getConnection().prepareStatement(stringBuilder.toString()).executeQuery();
            while (resultSet.next()) {
                return resultSet.getString(row);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Gets a stringlist from a table
     * @param tablename Requires a table name
     * @param row Requires the row of target stringlist
     * @param key (Optional) key
     * @return Returns the stringlist from the table
     */
    public List<String> getStringListFromTable(String tablename, String row, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("SELECT " + row + " FROM " + tablename + " ");
        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        try {
            ResultSet resultSet = database.getConnection().prepareStatement(stringBuilder.toString()).executeQuery();
            List<String> results = Lists.newCopyOnWriteArrayList();
            while (resultSet.next()) {
                results.add(resultSet.getString(row));
            }
            return results;
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Gets an integer from a table
     * @param tablename Requires a table name
     * @param row Requires the row of target integer
     * @param key (Optional) key
     * @return Returns the integer from the table
     */
    public int getIntegerFromTable(String tablename, String row, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("SELECT " + row + " FROM " + tablename + " ");
        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        try {
            ResultSet resultSet = database.getConnection().prepareStatement(stringBuilder.toString()).executeQuery();
            while (resultSet.next()) {
                return resultSet.getInt(row);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return 0;
    }

    /**
     * Gets a ResultSet
     * @param tablename Requires a tablename
     * @param key (Optional)key
     * @return Returns the ResultSet from the database
     */
    public ResultSet getResultSet(String tablename, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("SELECT * FROM " + tablename + " ");
        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        try {
            return database.getConnection().prepareStatement(stringBuilder.toString()).executeQuery();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Gets a ResultSet from the database
     * @param query Requires a query String
     * @return Returns the ResultSet from the database
     */
    public ResultSet getResultSet(String query) {
        try {
            return database.getConnection().prepareStatement(query).executeQuery();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * Inserts one or more things into the database
     * @param tablename Requires a tablename
     * @param insert Requires one or more Insert objects
     */
    public void insertIntoTable(String tablename, Insert... insert) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("INSERT INTO " + tablename + " (");

        int length = insert.length;

        for(Insert inserts : insert) {
            if(length != 1) {
                stringBuilder.append(inserts.getRow() + ", ");
                length--;
            } else {
                stringBuilder.append(inserts.getRow());
            }
        }

        stringBuilder.append(") VALUES (");
        length = insert.length;

        for(Insert inserts : insert) {
            if(length != 1) {
                stringBuilder.append("'" + inserts.getValue() + "', ");
                length--;
            } else {
                stringBuilder.append("'" + inserts.getValue() + "'");
            }
        }
        stringBuilder.append(");");
        executeCommand(stringBuilder.toString());
    }

    /**
     * Updates a value in a table
     * @param tablename Requires a tablename
     * @param row Requires the name of the row which should be updated
     * @param newValue Requires the new value
     * @param key (Optional) key(s)
     */
    public void updateValue(String tablename, String row, String newValue, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("UPDATE " + tablename + " SET " + row + " = '" + newValue + "' ");

        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        executeCommand(stringBuilder.toString());
    }

    /**
     * Updates one or more values in a table
     * @param tablename Requires the tablename
     * @param key Requires a key
     * @param insert Requires the new values as Insert objects
     */
    public void updateValues(String tablename, Key key, Insert... insert) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("UPDATE " + tablename + " SET ");

        int length = insert.length;
        if(length == 0) return;
        for(Insert inserts : insert) {
            if(length == 1) {
                stringBuilder.append(inserts.getRow());
                stringBuilder.append(" = '");
                stringBuilder.append(inserts.getValue());
                stringBuilder.append("' ");
            } else {
                stringBuilder.append(inserts.getRow());
                stringBuilder.append(" = '");
                stringBuilder.append(inserts.getValue());
                stringBuilder.append("', ");
                length--;
            }
        }
        stringBuilder.append(" WHERE ");
        stringBuilder.append(key.getRow() + " = '");
        stringBuilder.append(key.getKeyWord() + "'");

        executeCommand(stringBuilder.toString());
    }

    /**
     * If no key is provided, the method checks if the table is empty
     * @param tablename Requires the tablename
     * @param key (Optional) key(s)
     * @return Returns if the table does contain an entry
     */
    public boolean containsEntry(String tablename, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("SELECT * FROM " + tablename + " ");
        int length = key.length;
        if(length != 0) {
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append("WHERE " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        try {
            return database.getConnection().prepareStatement(stringBuilder.toString()).executeQuery().next();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * Deletes an entry from a table
     * @param tablename Requires the tablename
     * @param key (Optional) key(s)
     */
    public void deleteFromTable(String tablename, Key... key) {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("DELETE FROM ");
        stringBuilder.append(tablename + " WHERE");

        if(key.length != 0) {
            int length = key.length;
            for(Key keys : key) {
                if(length == 1) {
                    stringBuilder.append(" " + keys.getRow() + " = '" + keys.getKeyWord() + "'");
                } else {
                    stringBuilder.append(" " + keys.getRow() + " = '" + keys.getKeyWord() + "' AND ");
                    length--;
                }
            }
        }
        executeCommand(stringBuilder.toString());
    }

}
